libqtile/images.py

```diff
@@ -0,0 +1,205 @@
+        surf = cairocffi.ImageSurface.create_from_png(io.BytesIO(bytes_img))
+        return _SurfaceInfo(surf, 'png')
+    except (MemoryError, OSError):
+        pass
+    try:
+        surf, fmt = cairocffi.pixbuf.decode_to_image_surface(bytes_img)
+        return _SurfaceInfo(surf, fmt)
+    except cairocffi.pixbuf.ImageLoadingError:
+        pass
+    raise LoadingError("Couldn't load image!")
+
+def get_cairo_pattern(surface, width, height, theta=0.0):
+    """Return a SurfacePattern from an ImageSurface.
+
+    if width and height are not None scale the pattern
+    to be size width and height.
+
+    theta is in degrees ccw
+    """
+    EPSILON = 1.0e-6
+    from math import pi
+
+    if surface is None:
+        return None
+    pattern = cairocffi.SurfacePattern(surface)
+    pattern.set_filter(cairocffi.FILTER_BEST)
+    matrix = cairocffi.Matrix()
+    # TODO cleanup this function
+    tr_width, tr_height = None, None
+    if (width is not None) and (width != surface.get_width()):
+        tr_width = surface.get_width() / width
+    if (height is not None) and (height != surface.get_height()):
+        tr_height = surface.get_height() / height
+    if (tr_width is not None) or (tr_height is not None):
+        tr_width = tr_width if tr_width is not None else 1.0
+        tr_height = tr_height if tr_height is not None else 1.0
+        matrix.scale(tr_width, tr_height)
+
+    if abs(theta) > EPSILON:
+        theta_rad = pi / 180.0 * theta
+        mat_rot = cairocffi.Matrix.init_rotate(-theta_rad)
+        matrix = mat_rot.multiply(matrix)
+
+    pattern.set_matrix(matrix)
+    return pattern
+
+
+class Img(object):
+    """Img is a class which creates & manipulates cairo SurfacePatterns from an image
+
+    There are two constructors Img(...) and Img.from_path(...)
+
+    The cairo surface pattern is at img.pattern.
+    Changing any of the attributes width, height, or theta will update the pattern.
+
+    - width :: pattern width in pixels
+    - height :: pattern height in pixels
+    - theta :: rotation of pattern counter clockwise in degrees
+    Pattern is first stretched, then rotated.
+
+    - lock_aspect_ratio :: maintain aspect ratio when changing height or width
+    """
+    def __init__(self, bytes_img, name='', path='', lock_aspect_ratio=False):
+        self.lock_aspect_ratio = lock_aspect_ratio
+        self.bytes_img = bytes_img
+        self.name = name
+        self.path = path
+
+    @classmethod
+    def from_path(cls, image_path):
+        "Create an Img instance from image_path"
+        with open(image_path, 'rb') as fobj:
+            bytes_img = fobj.read()
+        name = os.path.basename(image_path)
+        name, file_type = os.path.splitext(name)
+        file_type = file_type.lstrip('.')
+        return cls(bytes_img, name=name, path=image_path)
+
+    @property
+    def file_type(self):
+        try:
+            return self._file_type
+        except AttributeError:
+            self.surface
```

Comment:
This seems like a bad code smell to have all these properties with side effects.